### PR TITLE
chore: adding script to collect integ test coverage

### PR DESCRIPTION
**Pull request by bombguy**
## Problem
We do not have ability to measure our current test coverage of our integration test suite. 

## Solution

- added `@cypress/code-coverage` & `nyc` dependency to measure test coverage during cypress run.
- updated `integ-setup` script to instrument the build artifacts of our sub-repos
- added `generateCoverageSummary` script to summarize the test coverage info and check if it meets predefined threshold.

## Additional Notes
Sample output of the generateCoverageSummary script
```
--------------------------------------------|---------|----------|---------|---------|----------------------------------------------------
File                                        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                  
--------------------------------------------|---------|----------|---------|---------|----------------------------------------------------
All files                                   |   74.32 |    57.13 |   73.56 |   73.53 |                                                    
 codegen-ui                                 |   77.15 |     54.2 |   81.81 |   76.62 |                                                    
  index.ts                                  |     100 |      100 |     100 |     100 |                                                    
  lib                                       |   76.79 |     54.2 |   81.81 |   76.23 |                                                    
   check-support.ts                         |   51.51 |       52 |   77.77 |   40.74 | 45-77                                              
   common-component-renderer.ts             |     100 |      100 |     100 |     100 |                                                    
   component-renderer-base.ts               |     100 |      100 |     100 |     100 |                                                    
   errors                                   |      64 |        0 |      60 |   60.86 |                                                    
    error-transformer.ts                    |   58.82 |        0 |   66.66 |   53.33 | 19-29,40                                           
    error-types.ts                          |   66.66 |      100 |      50 |   66.66 | 25-26                                              
    index.ts                                |     100 |      100 |     100 |     100 |                                                    
   framework-output-manager.ts              |     100 |      100 |     100 |     100 |                                                    
   generate-form-definition                 |   80.26 |    48.17 |   84.37 |    80.1 |                                                    
    generate-form-definition.ts             |   93.75 |    77.77 |     100 |   93.75 | 59                                                 
    helpers                                 |    79.2 |    47.69 |   81.81 |   79.06 |                                                    
     defaults.ts                            |     100 |      100 |     100 |     100 |                                                    
     field-type-map.ts                      |     100 |      100 |     100 |     100 |                                                    
     form-field.ts                          |   63.72 |    39.33 |   58.33 |   63.72 | 61-73,87-113,154,191,271-355,373,404               
     index.ts                               |     100 |      100 |     100 |     100 |                                                    
     map-cta.ts                             |    87.5 |    69.44 |     100 |    87.5 | 60,74,84                                           
     map-element.ts                         |    92.3 |    83.33 |     100 |    92.3 | 42                                                 
     map-styles.ts                          |     100 |    44.44 |     100 |     100 | 22-30                                              
     mapper-utils.ts                        |   84.61 |      100 |      75 |      90 | 25                                                 
     model-fields-configs.ts                |   96.93 |    83.76 |     100 |    96.9 | 106,187,321                                        
     position.ts                            |   65.62 |       60 |    62.5 |   63.93 | 34,38-39,45,75,78,83,86,97,116-137                 
     sectional-element.ts                   |   77.27 |    42.85 |     100 |   77.27 | 35-41,62,80,83                                     
    index.ts                                |     100 |      100 |     100 |     100 |                                                    
   generate-view-definition                 |   15.58 |        0 |    7.69 |   13.88 |                                                    
    generate-table-definition.ts            |   13.63 |        0 |       0 |      15 | 37-87                                              
    helpers                                 |    13.2 |        0 |       0 |   11.76 |                                                    
     index.ts                               |     100 |      100 |     100 |     100 |                                                    
     order-filter-columns.ts                |   11.53 |        0 |       0 |      10 | 27-44,48-52,57-82,88-99,103-129                    
    index.ts                                |     100 |      100 |     100 |     100 |                                                    
   generic-from-datastore.ts                |   91.17 |    77.96 |     100 |   91.17 | 39,53,59,71,100,106,117,122,127,163,176,179        
   renderer-helper.ts                       |     100 |    93.93 |     100 |     100 | 68                                                 
   required-dependency-provider.ts          |     100 |      100 |     100 |     100 |                                                    
   studio-node.ts                           |     100 |      100 |     100 |     100 |                                                    
   studio-template-renderer.ts              |      75 |      100 |      40 |   81.81 | 52-53                                              
   template-renderer-factory.ts             |   33.33 |      100 |       0 |   33.33 | 32-35                                              
   template-renderer.ts                     |      25 |        0 |       0 |      25 | 37-42,48-54,59-64                                  
   types                                    |    96.9 |      100 |   66.66 |   96.84 |                                                    
    bindings.ts                             |     100 |      100 |     100 |     100 |                                                    
    events.ts                               |     100 |      100 |     100 |     100 |                                                    
    form                                    |    93.1 |      100 |   33.33 |   92.59 |                                                    
     form-validation.ts                     |      92 |      100 |   33.33 |    91.3 | 55,59                                              
     index.ts                               |     100 |      100 |     100 |     100 |                                                    
    index.ts                                |     100 |      100 |     100 |     100 |                                                    
    relational-operator.ts                  |   66.66 |      100 |       0 |   66.66 | 24                                                 
    string-format.ts                        |     100 |      100 |     100 |     100 |                                                    
    view                                    |     100 |      100 |     100 |     100 |                                                    
     defaults.ts                            |     100 |      100 |     100 |     100 |                                                    
     index.ts                               |     100 |      100 |     100 |     100 |                                                    
   utils                                    |      75 |    71.36 |   90.47 |   73.68 |                                                    
    breakpoint-utils.ts                     |     100 |       75 |     100 |     100 | 43-44                                              
    component-metadata.ts                   |      94 |    91.04 |     100 |      94 | 166,189,259                                        
    component-tree.ts                       |   91.66 |    83.33 |     100 |    90.9 | 40                                                 
    form-component-metadata.ts              |   93.47 |      100 |    87.5 |   92.68 | 30,55-56                                           
    form-to-component                       |    85.5 |    68.29 |   94.11 |   84.37 |                                                    
     helpers                                |   79.54 |    69.23 |     100 |   78.57 |                                                    
      map-cta-buttons.ts                    |   75.75 |    58.33 |     100 |   74.19 | 35,87-88,93-98                                     
      map-element-children.ts               |    90.9 |    86.66 |     100 |    90.9 | 47                                                 
     index.ts                               |     100 |      100 |     100 |     100 |                                                    
     map-form-definition-to-component.ts    |   95.65 |       50 |   85.71 |   95.23 | 42                                                 
    form-utils.ts                           |   71.42 |    61.53 |     100 |   66.66 | 22,27                                              
    index.ts                                |     100 |      100 |     100 |     100 |                                                    
    reserved-words.ts                       |     100 |      100 |     100 |     100 |                                                    
    state-reference-metadata.ts             |     100 |     92.1 |     100 |     100 | 48,60                                              
    string-formatter.ts                     |   10.52 |        0 |       0 |   10.52 | 58-89,94-124,132-153,157-167                       
   validation-helper.ts                     |    92.5 |       65 |     100 |   91.66 | 70,74,78,125,160,230                               
 codegen-ui-react                           |   73.26 |    59.62 |   70.12 |   72.38 |                                                    
  index.ts                                  |     100 |      100 |     100 |     100 |                                                    
  lib                                       |   73.26 |    59.62 |   70.12 |   72.37 |                                                    
   amplify-ui-renderers                     |   77.64 |    64.64 |    92.3 |   77.08 |                                                    
    amplify-form-renderer.ts                |   36.92 |    22.64 |     100 |   34.92 | ...160-168,194,210-330,346-354,370-434,458-482,499 
    amplify-renderer.ts                     |     100 |      100 |     100 |     100 |                                                    
    amplify-view-renderer.ts                |      75 |       40 |     100 |      75 | 27,38-43                                           
    collection.ts                           |   97.29 |     73.8 |     100 |   97.22 | 83                                                 
    customComponent.ts                      |      90 |       75 |     100 |      90 | 31                                                 
    form.ts                                 |   92.15 |    71.42 |     100 |      92 | 62,101,118,201                                     
    sampleCodeRenderer.ts                   |      50 |      100 |   33.33 |      50 | 62-78                                              
   forms                                    |   75.11 |    63.29 |   73.82 |   73.52 |                                                    
    form-renderer-helper                    |      75 |     62.7 |   73.98 |   73.16 |                                                    
     all-props.ts                           |   98.18 |    95.08 |     100 |   98.14 | 348                                                
     bidirectional-relationship.ts          |   42.85 |     23.8 |   42.85 |   42.85 | 59,67-69,95-210,254-500                            
     cta-props.ts                           |   86.88 |    63.63 |     100 |    86.2 | 244-257,271,298-299,382                            
     event-handler-props.ts                 |   96.66 |    91.66 |     100 |   96.55 | 102,109,378                                        
     event-targets.ts                       |   87.87 |    84.37 |     100 |    87.5 | 90-91,153-154                                      
     form-state.ts                          |   86.47 |    78.43 |   79.41 |   86.01 | 104,161,173,335,440,498,520,536,551-553,574-602    
     index.ts                               |     100 |      100 |     100 |     100 |                                                    
     invalid-variable-helpers.ts            |    87.5 |       50 |     100 |    87.5 | 26                                                 
     label-decorator.ts                     |     100 |     87.5 |     100 |     100 | 105                                                
     map-from-fieldConfigs.ts               |   96.66 |       75 |   66.66 |   96.55 | 98                                                 
     model-fields.ts                        |     100 |    91.66 |     100 |     100 | 35                                                 
     model-values.ts                        |   73.42 |    54.83 |      84 |   73.38 | ...328,331,431-491,500,553,557-558,564-601,608,648 
     parse-fields.ts                        |   68.42 |    57.14 |   42.85 |    64.7 | 31,41,77,88,126-127,130-137                        
     relationship.ts                        |   32.74 |    26.66 |      40 |   29.62 | 62-203,214-1002,1110-1236,1279-1754                
     render-array-field.ts                  |     100 |    95.65 |     100 |     100 | 179                                                
     render-checkers.ts                     |     100 |      100 |     100 |     100 |                                                    
     static-props.ts                        |     100 |      100 |     100 |     100 |                                                    
     type-helper.ts                         |   22.52 |        0 |       0 |   17.47 | ...107-111,115-122,161-190,283-457,465-511,528-558 
     typescript-type-map.ts                 |     100 |      100 |     100 |     100 |                                                    
     validation.ts                          |   91.66 |    84.61 |   85.71 |    91.3 | 33-37                                              
     value-props.ts                         |   94.28 |    94.91 |     100 |   93.75 | 65,69,89,193                                       
    index.ts                                |     100 |      100 |     100 |     100 |                                                    
    react-form-renderer.ts                  |   75.13 |       68 |   68.75 |   75.13 | 203-244,308-339,388,436,467,485-488,525,627,634    
   helpers                                  |   69.23 |    37.93 |   71.42 |   64.44 |                                                    
    index.ts                                |   69.23 |    37.93 |   71.42 |   64.44 | 29,87,97-114,122,143                               
   imports                                  |      87 |    83.09 |      70 |    87.5 |                                                    
    import-collection.ts                    |   79.68 |    82.08 |   66.66 |      80 | 80,88-90,113-122                                   
    import-mapping.ts                       |     100 |      100 |     100 |     100 |                                                    
    index.ts                                |     100 |      100 |     100 |     100 |                                                    
   index.ts                                 |     100 |      100 |     100 |     100 |                                                    
   keywords.ts                              |     100 |      100 |     100 |     100 |                                                    
   name-replacements.ts                     |     100 |      100 |     100 |     100 |                                                    
   primitive.ts                             |   98.41 |      100 |   83.33 |   98.41 | 152                                                
   react-component-render-helper.ts         |   79.16 |    65.55 |   92.45 |   79.16 | ...438,446-452,466,472-482,506,544-569,609-620,634 
   react-component-renderer.ts              |     100 |     92.3 |     100 |     100 | 94-95,132,210                                      
   react-expander-renderer.ts               |   71.42 |       40 |     100 |   73.17 | 88,93,119-133,188-191                              
   react-index-studio-template-renderer.ts  |      95 |      100 |   85.71 |      95 | 62                                                 
   react-output-manager.ts                  |   33.33 |        0 |       0 |   33.33 | 22-42                                              
   react-render-config.ts                   |   82.35 |       50 |     100 |   78.57 | 34,38,50                                           
   react-required-dependency-provider.ts    |   66.66 |      100 |       0 |   66.66 | 24                                                 
   react-studio-template-renderer-helper.ts |   80.51 |    66.66 |   91.66 |   79.72 | 92-124,196,217,226,277                             
   react-studio-template-renderer.ts        |   63.88 |    59.34 |   61.53 |   64.74 | ...336-611,1099-1144,1329,1474,1519-1526,1539-1541 
   react-table-renderer-helper.ts           |    52.5 |     9.52 |      30 |   40.74 | 28-31,101-103,112-132                              
   react-table-renderer.ts                  |   10.63 |        0 |       0 |   10.63 | 34-314                                             
   react-theme-studio-template-renderer.ts  |   63.93 |    43.33 |   72.22 |   63.93 | 91-93,105-108,117,166,193-227,243                  
   react-utils-studio-template-renderer.ts  |    28.2 |        0 |       0 |    28.2 | 39-100                                             
   utils                                    |    43.6 |     6.06 |   17.85 |   36.75 |                                                    
    constants.ts                            |     100 |      100 |     100 |     100 |                                                    
    forms                                   |      40 |     6.59 |   13.04 |   33.33 |                                                    
     array-field-component.ts               |     100 |      100 |     100 |     100 |                                                    
     layout-helpers.ts                      |     100 |       75 |     100 |     100 | 22                                                 
     validation-helpers.ts                  |   65.21 |      100 |       0 |   55.55 | 29,51,85,270,349,653,1105,1423                     
     validation.ts                          |   13.79 |        0 |       0 |    9.09 | 19-39,43-44,48-156,165                             
     value-mappers.ts                       |   46.66 |        0 |       0 |   38.46 | 27-30,40-54                                        
    generate-react-hooks.ts                 |     100 |      100 |     100 |     100 |                                                    
    json-path-fetch.ts                      |   38.46 |        0 |       0 |   27.27 | 29-38,47                                           
    string-formatter.ts                     |      75 |      100 |       0 |     100 |                                                    
   views                                    |    63.2 |    51.11 |   42.85 |    63.2 |                                                    
    react-view-renderer.ts                  |    63.2 |    51.11 |   42.85 |    63.2 | 105-110,115,162-197,274,281-287,323-324,436-483    
   workflow                                 |   91.84 |    78.57 |   91.48 |   91.71 |                                                    
    action.ts                               |   96.92 |     90.9 |     100 |   96.92 | 64,194                                             
    events.ts                               |    92.3 |       80 |     100 |    92.3 | 64,114                                             
    index.ts                                |     100 |      100 |     100 |     100 |                                                    
    mutation.ts                             |    87.8 |    72.22 |   85.71 |    87.5 | 140-154                                            
    utils.ts                                |    87.5 |       50 |     100 |   85.71 | 26                                                 
--------------------------------------------|---------|----------|---------|---------|----------------------------------------------------
CoverageSummary {
  data: {
    lines: { total: 4244, covered: 3121, skipped: 0, pct: 73.53 },
    statements: { total: 4455, covered: 3311, skipped: 0, pct: 74.32 },
    functions: { total: 821, covered: 604, skipped: 0, pct: 73.56 },
    branches: { total: 3695, covered: 2111, skipped: 0, pct: 57.13 },
    branchesTrue: { total: 0, covered: 0, skipped: 0, pct: 'Unknown' }
  }
}
Failed to meet BRANCHES threshold: 57.13% < 58%
```

### Ticket

GitHub issue _____

### Automated tests
- [ ] Unit tests added/updated
- [ ] E2E tests added/updated
- [ ] N/A - (provide a reason)
- [ ] deferred - (provide GitHub issue for tracking)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/test-generator/integration-test-templates/cypress.config.ts [hunk 1]
```diff
@@ -17,7 +17,16 @@ import { defineConfig } from 'cypress';
 
 export default defineConfig({
   e2e: {
-    supportFile: false,
+    setupNodeEvents(on, config) {
+      // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
+      require('@cypress/code-coverage/task')(on, config);
+
+      // include any other plugin code...
```
Comment (lyonsbp) [inline packages/test-generator/integration-test-templates/cypress.config.ts:24]: Can we remove this comment?

File: packages/test-generator/integration-test-templates/cypress/scripts/generateCoverageSummary.mjs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+import libCoverage from 'istanbul-lib-coverage';
+import libReport from 'istanbul-lib-report';
+import reports from 'istanbul-reports';
+
+import coverageJSON from '../../coverage/coverage-final.json' assert { type: 'json' };
```
Comment (lyonsbp) [inline packages/test-generator/integration-test-templates/cypress/scripts/generateCoverageSummary.mjs:20]: What is assert doing here?

File: packages/test-generator/integration-test-templates/cypress/scripts/generateCoverageSummary.mjs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+// call execute to synchronously create and write the report to disk
+report.execute(context);
+
+// print combined summary of integ test coverage
+console.log(summary);
+
+Object.entries(MIN_COVERAGE_THRESHOLDS).forEach(([coverage, threshold]) => {
+  if (summary[coverage].pct <= threshold) {
+    console.error(
+      '\x1b[31m%s\x1b[0m',
```
Comment (lyonsbp) [inline packages/test-generator/integration-test-templates/cypress/scripts/generateCoverageSummary.mjs:58]: What is this?

File: packages/test-generator/integration-test-templates/cypress.config.ts (hunk 1 above)
Comment (bombguy) [reply packages/test-generator/integration-test-templates/cypress.config.ts:24]: Sure I can remove this comment

File: packages/test-generator/integration-test-templates/cypress/scripts/generateCoverageSummary.mjs (hunk 3 above)
Comment (bombguy) [reply packages/test-generator/integration-test-templates/cypress/scripts/generateCoverageSummary.mjs:58]: Just added some color make it easier to tell when the coverage drops below the threshold. Here is an example from my local terminal.

Screenshot 2023-03-16 at 1 04 36 PM

File: packages/test-generator/integration-test-templates/cypress/scripts/generateCoverageSummary.mjs (hunk 2 above)
Comment (bombguy) [reply packages/test-generator/integration-test-templates/cypress/scripts/generateCoverageSummary.mjs:20]: This was added due to type error that that gets thrown when you import the json file.

Check the run without the assert
https://github.com/aws-amplify/amplify-codegen-ui/actions/runs/4438598291/jobs/7789873668#step:8:27

```
TypeError [ERR_IMPORT_ASSERTION_TYPE_MISSING]: Module "file:///home/runner/work/amplify-codegen-ui/amplify-codegen-ui/packages/integration-test/coverage/coverage-final.json" needs an import assertion of type "json"
    at new NodeError (node:internal/errors:399:5)
    at validateAssertions (node:internal/modules/esm/assert:82:15)
    at defaultLoad (node:internal/modules/esm/load:84:3)
    at nextLoad (node:internal/modules/esm/loader:163:28)
    at ESMLoader.load (node:internal/modules/esm/loader:605:26)
    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:457:22)
    at new ModuleJob (node:internal/modules/esm/module_job:64:26)
    at #createModuleJob (node:internal/modules/esm/loader:480:[17](https://github.com/aws-amplify/amplify-codegen-ui/actions/runs/4438598291/jobs/7789873668#step:8:18))
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:434:34)
    at async ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:[21](https://github.com/aws-amplify/amplify-codegen-ui/actions/runs/4438598291/jobs/7789873668#step:8:22)) {
  code: 'ERR_IMPORT_ASSERTION_TYPE_MISSING'
}
```

**Review by lyonsbp — APPROVED** [review]

**Review by Jshhhh — APPROVED** [review]

**Review by hein-j — APPROVED** [review]

**Pull request merged**